google/cloud/storage/examples/storage_bucket_samples.cc

```diff
@@ -47,6 +47,26 @@ void PrintUsage(int argc, char* argv[], std::string const& msg) {
   std::cerr << std::flush;
 }
 
+void ListBuckets(storage::Client client, int& argc, char* argv[]) {
+  if (argc < 2) {
+    throw Usage{"list-buckets <project-id>"};
+  }
+  auto project_id = ConsumeArg(argc, argv);
+  //! [list buckets] [START storage_list_buckets]
+  [](google::cloud::storage::Client client, std::string project_id) {
+    int count = 0;
+    for (auto&& meta : client.ListBuckets(project_id)) {
```

Comment:
Question, project id should be optional given credentials provided. A service account could have granted permission to other projects and that would be a use-case for providing the project_id explicitly.

Reply:
Hmmm, the credentials could be from an authorized user, which also has access to (potentially) many projects.  Is there another type of credential where the project is fixed?

Note that some of the other languages require a project ID when you create the client object, so their samples for this API can just assume it is the same project being used here.  But we do not ask for a project id in the `Client` constructor because most APIs do not need one.  Effectively that means we do not know what project id to use here.

We could change the client class, but that is a longer discussion I think?

Reply:
+1 to needing to change the client class. This would be feasible if we had a constructor for ClientOptions that allowed specifying a default project, similar to some other client libs. e.g. Python [1].

[1] https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/storage/google/cloud/storage/client.py#L61

Reply:
I'll add a discussion into meeting notes for Monday. Agreed, it would mirror what existing clients do by defining this property in the client init. Other clients have that option and `GOOGLE_CLOUD_PROJECT` to define the selected project.

Reply:
I cannot make the meeting on Monday, I will be in transit to SEA.  I am going to merge this as-is, we can change it later (and/or add an additional version, whatever we decide).